asf_heat_pump_affordability/notebooks/investigate_effect_of_room_number.py

```diff
@@ -0,0 +1,496 @@
+
+
+for f in ["NUMBER_HABITABLE_ROOMS", "TOTAL_FLOOR_AREA"]:
+    generate_plot_archetype(f)
+
+# ## Findings
+# There are differences in average size of home between archetypes:
+# - Flats are smallest on average (~50m2) with fewer rooms (2)
+# - Bungalows and semi-detached/terraced houses are larger (~90m2) with more rooms (4 and 5, respectively)
+# - Detached houses are largest on average (~180m2) with the most rooms on average (7)
+
+# ## 3. Compare coefficients of quantile regression models
+#
+# The code below aims to do the following:
+# 1. Create 2 series of quantile regression models. Each series contains one regression model for each percentile 10-90. One series of models will be trained using the 8 different archetypes as the only features (a_models), the other with the 8 archetypes + number of rooms as the features (ar_models).
+# 2. For each model within both series of models: calculate the difference between pre- and post-1950 coefficients for each of the 4 property types (detached house, semi-detached/terraced house, flat, bungalow). E.g. For Quantile 0.1 model: pre-1950 flats coef - post-1950 flats coef, etc. Calculate the confidence intervals for the difference as well. Ultimately, we get: the difference at each quantile between pre- and post-1950 installation cost for each archetype and the confidence interval for this difference.
+# 3. For each property type, compare the difference in pre- and post-1950 installation cost at each quantile between both model groups (a_models vs ar_models).
+#
+# Note: confidence intervals spanning zero indicate there is a chance the difference in coefficients for pre- and post-1950 archetype is zero and therefore that there may be no meaningful difference in cost of installation when splitting the archetype by pre- or post-1950 build year.
+
+archetypes4 = _data.archetype4.unique()
+X_y = _data[["archetype", "NUMBER_HABITABLE_ROOMS", "adjusted_cost"]]
+
+## Regex pattern to extract archetype names from summary table
+pattern = r"\[(.*?)\]"
+
+# ## Series 1: models with archetypes as only features
+
+# +
+a_model_pre_post_coefs = {
+    "quantile": [],
+    "archetype": [],
+    "pre_1950_coef": [],
+    "post_1950_coef": [],
+    "pre_1950_coef_lower": [],
+    "post_1950_coef_lower": [],
+    "pre_1950_coef_upper": [],
+    "post_1950_coef_upper": [],
+}
+
+for quantile in np.arange(0.1, 1, 0.1):
+    quantile = round(quantile, 1)
+    # Train model
+    archetype_model = sm.quantreg("adjusted_cost ~ archetype - 1", X_y).fit(quantile)
+
+    # Get summary table containing coefficients, p-values, and confidence intervals for the model
+    table = archetype_model.summary(alpha=math.sqrt(0.05)).tables[1]
+    _df = pd.read_csv(
+        StringIO(table.as_csv()), index_col=0
+    ).reset_index()  # convert statsmodels table to df
+    _df.columns = ["feature", "coef", "std err", "t", "P>|t|", "lower", "upper"]
+    _df["feature"] = _df["feature"].apply(lambda x: re.search(pattern, x).group(1))
```

Comment:
Again, this works, but pandas already builds in this functionality into its API, you could have tried something like this:
`archetype_model.params.index.str.extract(pattern)`

pandas enables this `str` attribute on most (all?) of its objects, in this case the index object of the model parameters Series object. You don't even need to import `re` because it's already the default for the `extract()` method.